llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp

```diff
@@ -1820,6 +1971,26 @@ Instruction *InstCombinerImpl::visitFDiv(BinaryOperator &I) {
     return R;
 
   Value *Op0 = I.getOperand(0), *Op1 = I.getOperand(1);
+
+  // Convert
+  // x = 1.0/sqrt(a)
+  // r1 = x * x;
+  // r2 = a * x;
+  //
+  // TO
+  //
+  // r1 = 1/a
+  // r2 = sqrt(a)
+  // x = r1 * r2
+  SmallSetVector<Instruction *, 2> R1, R2;
+  getFSqrtDivOptPattern(&I, R1, R2);
+  if (!(R1.empty() || R2.empty()) && isFSqrtDivToFMulLegal(&I, R1, R2)) {
+    CallInst *CI = (CallInst *)((&I)->getOperand(1));
+    Value *SqrtOp = CI->getArgOperand(0);
+    if (Value *D = convertFSqrtDivIntoFMul(CI, &I, R1, R2, SqrtOp, Builder))
+      return replaceInstUsesWith(I, D);
+  }
```

Comment:
Instead of rooting this at the div and having to track down uses, maybe this should be root in `visitFMul`? (matching the `a * x` case). Think that would simplify a lot of the logic.

Reply:
I dont think that will simplify any of the logic because first detect the div pattern and from div, detect the  second mul pattern. This makes it detecting 2 patterns at least.
Same thing is being done above while rooting it in fdiv

Reply:
Oh I see, I think this should probably be in aggressiveInstCombine. @nikic any thoughts.

Reply:
Just on account of the use walk, or for some other reason? If it's just the use walk, I think it's ok to keep it in InstCombine. I'm less concerned about things like this for FP folds.

Reply:
Yeah, guess my feeling has always been `InstCombine` folds should be rooted in the instruction they are transforming. Past that its more aggressive, but not a hill I'll die on or anything.